### PR TITLE
Master worker pattern #799

### PR DESCRIPTION
**Pull request by AnaghaSasikumar**
Master worker pattern implementation

### CODE REVIEW

**Comment by AnaghaSasikumar** [comment]
Hi, I made the changes required...let me know if there are more changes to be made.

File: master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/ArrayTransposeMasterWorker.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+public class ArrayTransposeMasterWorker extends MasterWorker {
+
+  public ArrayTransposeMasterWorker() {
+    super(4);
+  }
+
+  @Override
+  Master setMaster(int numOfWorkers) {
+    ArrayTransposeMaster master = new ArrayTransposeMaster(numOfWorkers);
+    return master;
```
Comment (iluwatar) [inline master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/ArrayTransposeMasterWorker.java]: Could be one-liner

File: master-worker-pattern/src/main/java/com/iluwatar/masterworker/Result.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ *The abstract Result class, which contains 1 public field containing result
+ *data.
+ * @param <T> T will be type of data.
+ */
+
+public abstract class Result<T> {
+  
+  public T data;
```
[the remark this answers is not in the document]
Comment (iluwatar) [reply master-worker-pattern/src/main/java/com/iluwatar/masterworker/Result.java]: Check usage of `final` in other classes too

File: master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/MasterWorker.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  public MasterWorker(int numOfWorkers) {
+    this.master = setMaster(numOfWorkers);
+  }
+
+  abstract Master setMaster(int numOfWorkers);
+
+  public Result getResult(Input input) {
+    this.master.doWork(input);
+    return master.getFinalResult();
```
Comment (iluwatar) [inline master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/MasterWorker.java]: Could be one-liner?

File: master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/systemmaster/ArrayTransposeMaster.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      for (int m = 0; m < work.length; m++) {
+        //m = row number, n = columns number
+        for (int n = 0; n < work[0].length; n++) {
+          resultData[m][columnsDone + n] = work[m][n];
+        }
+      }
+      columnsDone += work[0].length;
+    }
+    ArrayResult result = new ArrayResult(resultData);
+    return result;
```
Comment (iluwatar) [inline master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/systemmaster/ArrayTransposeMaster.java]: Could be one-liner

File: master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/systemworkers/ArrayTransposeWorker.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    int[][] resultData = new int[((ArrayInput) this.getReceivedData()).data[0].length]
+      [((ArrayInput) this.getReceivedData()).data.length];
+    for (int i = 0; i < ((ArrayInput) this.getReceivedData()).data.length; i++) {
+      for (int j = 0; j < ((ArrayInput) this.getReceivedData()).data[0].length; j++) {
+        //flipping element positions along diagonal
+        resultData[j][i] = ((ArrayInput) this.getReceivedData()).data[i][j]; 
+      }
+    }
+    ArrayResult result = new ArrayResult(resultData);
+    return result;
```
Comment (iluwatar) [inline master-worker-pattern/src/main/java/com/iluwatar/masterworker/system/systemworkers/ArrayTransposeWorker.java]: one-liner

**Comment by iluwatar** [comment]
@AnaghaSasikumar I posted my review comments. Overall looks very good. Just some minor improvements mentioned there.

**Comment by AnaghaSasikumar** [comment]
I made changes according to the review comments

**Comment by iluwatar** [comment]
Looks good @AnaghaSasikumar Thank you very much for this new pattern! :)

**Pull request merged**